# 01 Importing libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [4]:
# Create strings for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [5]:
# Import Orders-Products data after merging and deriving new variables
df_ords_prods_merge=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_new_vars.pkl'))

In [6]:
df_ords_prods_merge.shape

(32399732, 18)

In [7]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices    price_range_loc  \
0          0         Soda        77              7     9.0  Mid range product   
1          1         Soda        77              7     9.0  Mid range product   
2          1         Soda        77              7     9.0  Mid range product   
3          1         Soda        77              7     9.0  Mid range product   
4          1         Soda        77              7     9.0  Mid range product   

    busiest_day busiest_slowest_days busiest_period_of_day  
0  Regular busy         Regular busy        Average orders  
1  Regular busy         Slowest days        Average orders  
2  Regular busy         Slowest days           Most orders  
3    Least busy         Slowest days        Average orders  
4    Least busy         Slowest days           Most orders

# Step 2. Aggregated mean of the “order_number” column grouped by “department_id” 

In [9]:
# Define the mean value of 'order number' grouped by 'department id'
df_ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.667853
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# Step 3. Analyzing the result of agreegation of the order number mean value

# The output displays the mean values for 21 departments from 1 to 21 vs eight departments in the subset. The mean value differs for the departments presenting in both dataframes (entire one and subset), e.g. for the department ID ‘20’ the mean is equal to 17.599636 in the subset vs 16.473447 in the entire data. It is evident that the limited selection of rows affect the statistics of a variable.   

# Step 4. Creating a loyalty flag for customers

In [19]:
# Create a new column 'max_order' and calculate maximum orders per customer
df_ords_prods_merge['max_order']=df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Oksana Stepanova\AppData\Local\Temp\ipykernel_11700\2686850048.py:2: FutureWarning: The provided callable <function amax at 0x0000017E3C85E700> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order']=df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [20]:
df_ords_prods_merge[['user_id','max_order']].head()

user_id  max_order
0        1         10
1        1         10
2        1         10
3        1         10
4        1         10

In [21]:
# Create a new column 'loyalty_flag' and assign loyalty label 'Loyal customer' if order number >40
df_ords_prods_merge.loc[df_ords_prods_merge['max_order']>40, 'loyalty_flag']='Loyal customer'

In [22]:
# Create a new column 'loyalty_flag' and assign loyalty label 'Regular customer' if order nember <=40 and >10 
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order']<=40) & (df_ords_prods_merge['max_order']>10), 'loyalty_flag']='Regular customer'

In [24]:
# Create a new column 'loyalty_flag' and assign loyalty label 'New customer' if order nember <=10
df_ords_prods_merge.loc[df_ords_prods_merge['max_order']<=10, 'loyalty_flag']='New customer'

In [25]:
df_ords_prods_merge.shape

(32399732, 21)

In [26]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ... aisle_id  department_id  prices    price_range_loc  \
0          0  ...       77              7     9.0  Mid range product   
1          1  ...       77              7     9.0  Mid range product   
2          1  ...       77              7     9.0  Mid range product   
3          1  ...       77              7     9.0  Mid range product   
4          1  ...       77              7     9.0  Mid range product   

    busiest_day busiest_slowest_days busiest_period_of_day average_spend  \
0  Regular busy         Regular busy        Average orders      6.367797   
1  Regular busy         Slowest days        Average orders      6.367797   
2  Regular busy         Slowest days           Most orders      6.367797   
3    Least busy         Slowest days        Average orders      6.367797   
4    Least busy         Slowest days           Most orders      6.367797   

   max_order  loyalty_flag  
0         10  New customer  
1         10  New customer  
2         10  New customer  
3         10  New customer  
4         10  New customer  

[5 rows x 21 columns]

In [27]:
# Make a correction: remove a column 'average_spend' previously created on a wrong step
df_ords_prods_merge=df_ords_prods_merge.drop(columns=['average_spend'])

In [29]:
# Check for the dataframe after removeing the column 'average_spend'
df_ords_prods_merge.shape

(32399732, 20)

In [30]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered product_name  aisle_id  department_id  prices    price_range_loc  \
0          0         Soda        77              7     9.0  Mid range product   
1          1         Soda        77              7     9.0  Mid range product   
2          1         Soda        77              7     9.0  Mid range product   
3          1         Soda        77              7     9.0  Mid range product   
4          1         Soda        77              7     9.0  Mid range product   

    busiest_day busiest_slowest_days busiest_period_of_day  max_order  \
0  Regular busy         Regular busy        Average orders         10   
1  Regular busy         Slowest days        Average orders         10   
2  Regular busy         Slowest days           Most orders         10   
3    Least busy         Slowest days        Average orders         10   
4    Least busy         Slowest days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

# Step 5. Check for values on loyalty flag column

In [34]:
# Count the frequency of values on loyalty flag column
df_ords_prods_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15874128
Loyal customer      10282763
New customer         6242841
Name: count, dtype: int64

# Step 6. Creating a Spending flag for customers 

In [36]:
# Create a new column 'average_spending' and calculate the mean of the prices of products purchased by each customer
df_ords_prods_merge['average_spending']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Oksana Stepanova\AppData\Local\Temp\ipykernel_11700\3262871512.py:2: FutureWarning: The provided callable <function mean at 0x0000017E3C85F420> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['average_spending']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [37]:
# Check for the dataframe after adding a new column 'average_spending'
df_ords_prods_merge.shape

(32399732, 21)

In [38]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ... aisle_id  department_id  prices    price_range_loc  \
0          0  ...       77              7     9.0  Mid range product   
1          1  ...       77              7     9.0  Mid range product   
2          1  ...       77              7     9.0  Mid range product   
3          1  ...       77              7     9.0  Mid range product   
4          1  ...       77              7     9.0  Mid range product   

    busiest_day busiest_slowest_days busiest_period_of_day max_order  \
0  Regular busy         Regular busy        Average orders        10   
1  Regular busy         Slowest days        Average orders        10   
2  Regular busy         Slowest days           Most orders        10   
3    Least busy         Slowest days        Average orders        10   
4    Least busy         Slowest days           Most orders        10   

   loyalty_flag average_spending  
0  New customer         6.367797  
1  New customer         6.367797  
2  New customer         6.367797  
3  New customer         6.367797  
4  New customer         6.367797  

[5 rows x 21 columns]

In [39]:
# Create a new column 'spender_flag' and assign the label 'Low spender' if 'average_spending' <10
df_ords_prods_merge.loc[df_ords_prods_merge['average_spending']<10, 'spender_flag']='Low spender'

In [40]:
# Create a new column 'spender_flag' and assign the label 'High spender' if 'average_spending' >=10
df_ords_prods_merge.loc[df_ords_prods_merge['average_spending']>=10, 'spender_flag']='High spender'

In [41]:
# Check for the dataframe after adding a new column 'spender_flag'
df_ords_prods_merge.shape

(32399732, 22)

In [42]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ... department_id  prices    price_range_loc   busiest_day  \
0          0  ...             7     9.0  Mid range product  Regular busy   
1          1  ...             7     9.0  Mid range product  Regular busy   
2          1  ...             7     9.0  Mid range product  Regular busy   
3          1  ...             7     9.0  Mid range product    Least busy   
4          1  ...             7     9.0  Mid range product    Least busy   

  busiest_slowest_days busiest_period_of_day max_order  loyalty_flag  \
0         Regular busy        Average orders        10  New customer   
1         Slowest days        Average orders        10  New customer   
2         Slowest days           Most orders        10  New customer   
3         Slowest days        Average orders        10  New customer   
4         Slowest days           Most orders        10  New customer   

   average_spending spender_flag  
0          6.367797  Low spender  
1          6.367797  Low spender  
2          6.367797  Low spender  
3          6.367797  Low spender  
4          6.367797  Low spender  

[5 rows x 22 columns]

In [43]:
# Count the frequency of values on Spender flag column
df_ords_prods_merge['spender_flag'].value_counts(dropna=False)

spender_flag
Low spender     32280013
High spender      119719
Name: count, dtype: int64

# Step 7. Creating an order frequency flag for customers  

In [45]:
# Check for the values in 'days_since_prior_order' column
df_ords_prods_merge['days_since_prior_order'].describe()

count    3.032399e+07
mean     1.110395e+01
std      8.779012e+00
min      0.000000e+00
25%      5.000000e+00
50%      8.000000e+00
75%      1.500000e+01
max      3.000000e+01
Name: days_since_prior_order, dtype: float64

In [47]:
df_ords_prods_merge['days_since_prior_order'].isnull().sum()

2075745

# There are 2,075,745 NaNs values related to the first order placed by the client.

In [49]:
# Create a new column 'median_days_prior_order' and calculate the median of 'days_since_prior_order' per customer
df_ords_prods_merge['median_days_prior_order']=df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Oksana Stepanova\AppData\Local\Temp\ipykernel_11700\278422407.py:2: FutureWarning: The provided callable <function median at 0x0000017E3C9A6020> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['median_days_prior_order']=df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [50]:
# Check for the dataframe after adding a new column 'median_days_prior_order'
df_ords_prods_merge.shape

(32399732, 23)

In [51]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ... prices    price_range_loc   busiest_day  \
0          0  ...    9.0  Mid range product  Regular busy   
1          1  ...    9.0  Mid range product  Regular busy   
2          1  ...    9.0  Mid range product  Regular busy   
3          1  ...    9.0  Mid range product    Least busy   
4          1  ...    9.0  Mid range product    Least busy   

   busiest_slowest_days busiest_period_of_day max_order  loyalty_flag  \
0          Regular busy        Average orders        10  New customer   
1          Slowest days        Average orders        10  New customer   
2          Slowest days           Most orders        10  New customer   
3          Slowest days        Average orders        10  New customer   
4          Slowest days           Most orders        10  New customer   

  average_spending  spender_flag median_days_prior_order  
0         6.367797   Low spender                    20.5  
1         6.367797   Low spender                    20.5  
2         6.367797   Low spender                    20.5  
3         6.367797   Low spender                    20.5  
4         6.367797   Low spender                    20.5  

[5 rows x 23 columns]

In [52]:
# Create a new column 'order_frequency_flag' and assign the label 'Non-frequent customer' if the median >20
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_prior_order']>20, 'order_frequency_flag']='Non-frequent customer'

In [53]:
# Create a new column 'order_frequency_flag' and assign the label 'Regular customer' if the median >10 and <=20
df_ords_prods_merge.loc[(df_ords_prods_merge['median_days_prior_order']>10) & (df_ords_prods_merge['median_days_prior_order']<=20), 'order_frequency_flag']='Regular customer'

In [54]:
# Create a new column 'order_frequency_flag' and assign the label 'Frequent customer' if the median <=10
df_ords_prods_merge.loc[df_ords_prods_merge['median_days_prior_order']<=10, 'order_frequency_flag']='Frequent customer'

In [55]:
# Check for the dataframe after adding a new column 'order_frequency_flag'
df_ords_prods_merge.shape

(32399732, 24)

In [56]:
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ...    price_range_loc   busiest_day  busiest_slowest_days  \
0          0  ...  Mid range product  Regular busy          Regular busy   
1          1  ...  Mid range product  Regular busy          Slowest days   
2          1  ...  Mid range product  Regular busy          Slowest days   
3          1  ...  Mid range product    Least busy          Slowest days   
4          1  ...  Mid range product    Least busy          Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spending  \
0         Average orders        10  New customer         6.367797   
1         Average orders        10  New customer         6.367797   
2            Most orders        10  New customer         6.367797   
3         Average orders        10  New customer         6.367797   
4            Most orders        10  New customer         6.367797   

  spender_flag  median_days_prior_order   order_frequency_flag  
0  Low spender                     20.5  Non-frequent customer  
1  Low spender                     20.5  Non-frequent customer  
2  Low spender                     20.5  Non-frequent customer  
3  Low spender                     20.5  Non-frequent customer  
4  Low spender                     20.5  Non-frequent customer  

[5 rows x 24 columns]

In [58]:
# Count the frequency of values on Order frequency flag column
df_ords_prods_merge['order_frequency_flag'].value_counts(dropna=False)

order_frequency_flag
Frequent customer        21556644
Regular customer          7207340
Non-frequent customer     3635743
nan                             5
Name: count, dtype: int64

In [62]:
# Change row numbers settings of head() function
pd.options.display.max_rows=None

In [63]:
# Print the first 60 rows of selected columns on Orders-Products dataframe
df_ords_prods_merge[['user_id', 'loyalty_flag', 'spender_flag', 'order_frequency_flag']].head(60)

user_id      loyalty_flag spender_flag   order_frequency_flag
0         1      New customer  Low spender  Non-frequent customer
1         1      New customer  Low spender  Non-frequent customer
2         1      New customer  Low spender  Non-frequent customer
3         1      New customer  Low spender  Non-frequent customer
4         1      New customer  Low spender  Non-frequent customer
5         1      New customer  Low spender  Non-frequent customer
6         1      New customer  Low spender  Non-frequent customer
7         1      New customer  Low spender  Non-frequent customer
8         1      New customer  Low spender  Non-frequent customer
9         1      New customer  Low spender  Non-frequent customer
10       15  Regular customer  Low spender      Frequent customer
11       15  Regular customer  Low spender      Frequent customer
12       15  Regular customer  Low spender      Frequent customer
13       15  Regular customer  Low spender      Frequent customer
14       15  Regular customer  Low spender      Frequent customer
15       19      New customer  Low spender      Frequent customer
16       19      New customer  Low spender      Frequent customer
17       19      New customer  Low spender      Frequent customer
18       21  Regular customer  Low spender      Frequent customer
19       31  Regular customer  Low spender      Frequent customer
20       31  Regular customer  Low spender      Frequent customer
21       43  Regular customer  Low spender      Frequent customer
22       43  Regular customer  Low spender      Frequent customer
23       52  Regular customer  Low spender      Frequent customer
24       52  Regular customer  Low spender      Frequent customer
25       52  Regular customer  Low spender      Frequent customer
26       52  Regular customer  Low spender      Frequent customer
27       52  Regular customer  Low spender      Frequent customer
28       52  Regular customer  Low spender      Frequent customer
29       52  Regular customer  Low spender      Frequent customer
30       52  Regular customer  Low spender      Frequent customer
31       52  Regular customer  Low spender      Frequent customer
32       52  Regular customer  Low spender      Frequent customer
33       52  Regular customer  Low spender      Frequent customer
34       52  Regular customer  Low spender      Frequent customer
35       52  Regular customer  Low spender      Frequent customer
36       52  Regular customer  Low spender      Frequent customer
37       67  Regular customer  Low spender      Frequent customer
38       67  Regular customer  Low spender      Frequent customer
39       67  Regular customer  Low spender      Frequent customer
40       67  Regular customer  Low spender      Frequent customer
41       67  Regular customer  Low spender      Frequent customer
42       67  Regular customer  Low spender      Frequent customer
43       67  Regular customer  Low spender      Frequent customer
44       67  Regular customer  Low spender      Frequent customer
45       67  Regular customer  Low spender      Frequent customer
46       67  Regular customer  Low spender      Frequent customer
47       67  Regular customer  Low spender      Frequent customer
48       67  Regular customer  Low spender      Frequent customer
49       67  Regular customer  Low spender      Frequent customer
50       67  Regular customer  Low spender      Frequent customer
51       67  Regular customer  Low spender      Frequent customer
52       67  Regular customer  Low spender      Frequent customer
53       67  Regular customer  Low spender      Frequent customer
54       67  Regular customer  Low spender      Frequent customer
55       67  Regular customer  Low spender      Frequent customer
56       81      New customer  Low spender  Non-frequent customer
57       81      New customer  Low spender  Non-frequent customer
58       82  Regular customer  Low spender       Regular customer
59       82  Regular customer  Low spender  

# Step 9. Exporting data 

In [69]:
# Export dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

# Task revision

## 01 Importing libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02 Importing Data

In [5]:
# Create strings for path
path=r'C:\Users\Oksana Stepanova\OneDrive\Документи\Instacart Basket Analysis'

In [7]:
# Import Orders-Products data after grouping data and aggregating variables
df_ords_prods_agg=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_aggregated.pkl'))

In [8]:
df_ords_prods_agg.shape

(32399732, 24)

In [9]:
df_ords_prods_agg.head(5)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                    15.0        False         196                  1   
2                    21.0        False         196                  1   
3                    29.0        False         196                  1   
4                    28.0        False         196                  1   

   reordered  ...    price_range_loc   busiest_day  busiest_slowest_days  \
0          0  ...  Mid range product  Regular busy          Regular busy   
1          1  ...  Mid range product  Regular busy          Slowest days   
2          1  ...  Mid range product  Regular busy          Slowest days   
3          1  ...  Mid range product    Least busy          Slowest days   
4          1  ...  Mid range product    Least busy          Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_spending  \
0         Average orders        10  New customer         6.367797   
1         Average orders        10  New customer         6.367797   
2            Most orders        10  New customer         6.367797   
3         Average orders        10  New customer         6.367797   
4            Most orders        10  New customer         6.367797   

  spender_flag  median_days_prior_order   order_frequency_flag  
0  Low spender                     20.5  Non-frequent customer  
1  Low spender                     20.5  Non-frequent customer  
2  Low spender                     20.5  Non-frequent customer  
3  Low spender                     20.5  Non-frequent customer  
4  Low spender                     20.5  Non-frequent customer  

[5 rows x 24 columns]

In [10]:
df_ords_prods_agg.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'price_range_loc', 'busiest_day',
       'busiest_slowest_days', 'busiest_period_of_day', 'max_order',
       'loyalty_flag', 'average_spending', 'spender_flag',
       'median_days_prior_order', 'order_frequency_flag'],
      dtype='object')

## Step 5. The spending habits of Loyal Regular and New customers

In [14]:
# Check the mean, min and max product prices for each loyalty category
df_ords_prods_agg.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773575  1.0  25.0
New customer      7.801206  1.0  25.0
Regular customer  7.798262  1.0  25.0

In [15]:
df_ords_prods_agg['prices'].describe()

count    3.239973e+07
mean     7.790994e+00
std      4.241809e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

### Observations: The average price of products purchased by new customers is the highest among the three categories of customers ranked by their loyalty and equal to 7.80 USD. While the average price per product spent by loyal customers is the lowest, e.g. 7.77 USD. The minimum and maximum prices of products are the same for all categories of customers.         